src/mme-app/handlers/paging.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2003-2018, Great Software Laboratory Pvt. Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <string.h>
+
+#include "mme_app.h"
+#include "ue_table.h"
+#include "err_codes.h"
+#include "message_queues.h"
+#include "ipc_api.h"
+#include "paging_info.h"
+
+/************************************************************************
+Paging handler.
+Flow : 36.413 - release 15 - Section 8.5
+		DDN Request : SGW-->MME 
+--->		Paging MME --> eNB
+**************************************************************************/
+
+/****Globals and externs ***/
+
+extern int g_mme_hdlr_status;
+
+static int g_Q_paging_fd;
+static int g_Q_DDN_fd;
+
+/*Making global just to avoid stack passing*/
+static char DDN_buf[10];//[S6A_DDN_BUF_SIZE];
+/*tmp defined*/
+#define S6A_DDN_BUF_SIZE 10
+
+extern uint32_t paging_counter;
+/****Global and externs end***/
+
+/**
+Initialize the stage settings, Q,
+destination communication etc.
+*/
+static void
+init_stage()
+{
+	log_msg(LOG_INFO, "DDN reader complete: waiting\n");
+	if ((g_Q_DDN_fd  = open_ipc_channel(
+					S6A_DDN_QUEUE, IPC_READ)) == -1){
+		log_msg(LOG_ERROR, "Error in opening reader - DDN"
+				"S6A IPC channel.\n");
+		pthread_exit(NULL);
+	}
+	log_msg(LOG_INFO, "DDN reader complete: Connected\n");
+	g_Q_paging_fd = open_ipc_channel(S1AP_PAGING_QUEUE, IPC_WRITE);
+	if (g_Q_paging_fd == -1) {
+		log_msg(LOG_ERROR, "Error in opening Writer IPC channel:s1ap Paging\n");
+		pthread_exit(NULL);
+	}
+	log_msg(LOG_INFO, "s1ap Paging writer: Connected\n");
+
+	return;
+}
+
+/**
+* Read next message from stage Q for processing.
+*/
+static int
+read_next_msg()
+{
+	int bytes_read=0;
+
+	memset(DDN_buf, 0, S6A_DDN_BUF_SIZE);
+	while (bytes_read < S6A_DDN_BUF_SIZE) {//TODO : Recheck condition
+		if ((bytes_read = read_ipc_channel(
+						g_Q_DDN_fd,
+						DDN_buf,
+						S6A_DDN_BUF_SIZE)) == -1) {
+			log_msg(LOG_ERROR, "Error in reading \n");
+			/* TODO : Add proper error handling */
+		}
+		log_msg(LOG_INFO, "DDN Msg Received - Len: %d\n",
+			bytes_read);
+	}
+
+	return bytes_read;
+}
+
+/**
+* Stage specific message processing.
+*/
+static int
+DDN_processing()
+{
+#ifdef UNCOMMENT_WHEN_S11_DDN_IS_DONE
+
+	struct DDN_Q_msg *ddn_info = (struct DDN_Q_msg *)buf;
```

Comment:
Can you please define DDN_Q_msg structure. then we can ask other team (working on s11) to fill the structure for you. That way your code will be clean compiled code. We can also inject DDN from s11-app through timer, just to check if we our pipe communication and s1ap encoding is correct or not.

Reply:
Intentionally had left as was not aware of what other team is up to. I'll add the structures, while integration those can be modified.